wfcGen.py

```diff
@@ -164,85 +166,97 @@ def __init__(self, sx, sy, cells, tileSize=3):
 
     def setRandom(self):
         self.grid = []
+        self.entropySave = {}
         for y in range(self.sy):
             self.grid.append([])
             for x in range(self.sx):
                 self.grid[y].append(Cell([x, y]))
-        self.grid[random.randint(0, self.sy-1)][random.randint(0, self.sx-1)].setCell(random.randint(0, len(self.cells)-1), self.cells)
-        self.getFinalImage()
-
-    def loadEntropy(self):
-        self.entropySave = {}
-        for y in range(0, self.sy):
-            for x in range(0, self.sx):
-                self.grid[y][x].loadEntropy(self.grid)
-                if not self.grid[y][x].collapsed:
-                    entropy = self.grid[y][x].getEntropy(self.cells)
-                    self.entropySave[entropy] = self.entropySave.get(entropy, []) + [[x, y]]
+        rand = [random.randint(0, self.sy-1), random.randint(0, self.sx-1)]
+        self.grid[rand[0]][rand[1]].setCell(random.randint(0, len(self.cells)-1), self.cells)
+        self.loadPosEntropy([rand[1], rand[0]])
+        self.loadEntropyAroundPos([rand[1], rand[0]])
+
+    def loadPosEntropy(self, pos):
+        if (pos[0], pos[1]) in list(self.invertedEntropy.keys()) and pos in self.entropySave[self.invertedEntropy[(pos[0], pos[1])]]:
+            self.entropySave[self.invertedEntropy[(pos[0], pos[1])]].remove(pos)
+            if self.entropySave[self.invertedEntropy[(pos[0], pos[1])]] == []:
+                self.entropySave.pop(self.invertedEntropy[(pos[0], pos[1])], None)
+        self.grid[pos[1]][pos[0]].loadEntropy(self.grid)
+        entropy = self.grid[pos[1]][pos[0]].getEntropy(self.cells)
+        if entropy != -1:
+            self.entropySave[entropy] = self.entropySave.get(entropy, []) + [pos]
+            self.invertedEntropy[(pos[0], pos[1])] = entropy
+        else:
+            self.invertedEntropy.pop((pos[0], pos[1]), None)
+    
+    def loadEntropyAroundPos(self, pos):
+        for x in [-1, 0, 1]:
+            for y in [-1, 0, 1]:
+                if abs(x) == abs(y):
+                    continue
+                if (pos[1]+y)<len(self.grid) and (pos[0]+x)<len(self.grid) and (pos[1]+y)>=0 and (pos[0]+x)>=0:
+                    self.loadPosEntropy([pos[0]+x, pos[1]+y])
 
     def collapse(self):
-        self.loadEntropy()
         sortedEntropy = sorted(self.entropySave.keys())
         if len(sortedEntropy) == 0:
                 return True
         while min(sortedEntropy) < 1:
             sortedEntropy.remove(min(sortedEntropy))
             if len(sortedEntropy) == 0:
                 return True
-        y = self.entropySave[sortedEntropy[0]][0][1]
-        x = self.entropySave[sortedEntropy[0]][0][0]
+        entropyPos = random.choice(self.entropySave[sortedEntropy[0]])
+        y = entropyPos[1]
+        x = entropyPos[0]

```

Comment:
Is it `x, y = entropyPos`?

Let me remind you, unpacking a tuple instead of indexing its members gives the advantage of a type check that ensures the number of members:

```python
def unpack1(coords3):
    x, y, z = coords3

def unpack2(coords3):
    x = coords3[0]
    y = coords3[1]
    z = coords3[2]

if __name__ == '__main__':
    # Successful either way:
    pos = (1, 2, 3)
    unpack1(pos)
    unpack2(pos)

    # Exception raised either way:
    pos = (1, 2)
    unpack1(pos)  # Raises `TypeError`
    unpack2(pos)  # Raises `IndexError`

    # Only one exception!
    pos = (1, 2, 3, 4)
    unpack1(pos)  # Raises `TypeError`, so you know something is wrong
    unpack2(pos)  # Silently ignores the problem, you don't get to know anything was wrong
```

In cases when you really need to extract only a few members of a tuple, there is an option to do this with unpacking too:
```python
# Simple case: want to extract just `1` and `3`
redundant = (1, 2, 3, 4)
x, _, y, _ = redundant  # Yay!

# Nested case: want to extract `2` and `4`
redundant = (1, (2, 3), 4, (5,))
_, (x, _), y, (_,) = redundant  # Yay!
```